machinelearning/tensorflow.md

```diff
@@ -0,0 +1,415 @@
+```
+
+**Using Docker hub container**
+
+```
+# pull the tensorflow docker container with onednn-acl optimizations enabled
+docker pull armswdev/tensorflow-arm-neoverse
+
+# launch the docker image
+docker run -it --rm -v /home/ubuntu/:/hostfs armswdev/tensorflow-arm-neoverse
+```
+
+**Using AWS DLAMI**
+
+To launch a Graviton instance with a DLAMI via EC2 console:
+
+1. Click on "Community AMIs"
+2. Filter AMIs by clicking Ubuntu and Arm (64-bit) check boxes
+3. Search for "Deep Learning" in search field to list the current Ubuntu based DLAMIs"
+4. Select the AMI: "Deep Learning AMI Graviton Tensorflow <version>(Ubuntu 20.04) <yyyy/mm/dd>"
+
+Here is the awscli snippet to find the latest AMI and launch it:
+```
+aws ec2 describe-images --region us-west-2 --filters  Name=architecture,Values=arm64 Name=name,Values="*Deep Learning AMI*" Name=owner-alias,Values=amazon  --query 'Images[] | sort_by(@, &CreationDate)[-1] | ImageId'
+
+aws ec2 run-instances --region us-west-2 --image-id <ami-id from the above command>  --instance-type c7g.4xlarge --count 1  --key-name <key name> --subnet-id <subnet_id> --block-device-mappings 'DeviceName=/dev/sda1,Ebs={VolumeSize=256}'
+```
+Once Graviton instance is launched with the above DLAMI, the platform is ready for building Tensorflow based ML inference applications.The DLAMI comes with the additional libraries preinstalled, eg: rust compilers, transformers etc, to enable transformers based (eg: bert) inferencing applications.
+
+**Using Tensorflow Serving with AWS DLC**
+
+```
+# login and pull the AWS DLC for tensorflow
+aws ecr get-login-password --region us-west-2 | docker login --username AWS --password-stdin 763104351884.dkr.ecr.us-west-2.amazonaws.com
+
+docker pull 763104351884.dkr.ecr.us-west-2.amazonaws.com/tensorflow-inference-graviton:2.7.0-cpu-py38-ubuntu20.04-e3-v1.0
+
+#sample command to launch the tensorflow serving api with resnet50 model
+docker run -p 8501:8501 --name tfserving_resnet --mount type=bind,source=/tmp/resnet,target=/models/resnet -e MODEL_NAME=resnet -t 763104351884.dkr.ecr.us-west-2.amazonaws.com/tensorflow-inference-graviton:2.7.0-cpu-py38-ubuntu20.04-e3-v1.0
+
+```
+Note: The above Tensorflow DLC supports the default Eigen backend. Please refer to the **Enable Tensorflow serving with onednn+acl backend** section below to build Tensorflow serving docker image with onednn+acl optimizations for better performance.
+
+
+# Runtime configurations for optimal performance
+Once the Tensorflow setup is ready, enable the below runtime configurations to achieve the best performance.
+
+```
+# The default runtime backend for tensoflow is eigen, but typically onednn+acl provides better performance and this can be enabled by setting the below TF environment variable
+export TF_ENABLE_OENDNN_OPTS=1
+
+# Graviton3 (eg: c7g instances) supports BF16 format for ML acceleration. This can be enabled in oneDNN by setting the below environment variable
+export DNNL_DEFAULT_FPMATH_MODE=BF16
```

Comment:
yes, your statement is more accurate,  I will update. I went without checking the device caps, because onednn/acl does the check and ignores BF16 setting if the device doesn't support it.